portable/ThirdParty/GCC/Posix/port.c

```diff
@@ -231,15 +255,26 @@ BaseType_t xPortStartScheduler( void )
         sigwait( &xSignals, &iSignal );
     }
 
-    /* Cancel the Idle task and free its resources */
-    #if ( INCLUDE_xTaskGetIdleTaskHandle == 1 )
-        vPortCancelThread( xTaskGetIdleTaskHandle() );
-    #endif
-
-    #if ( configUSE_TIMERS == 1 )
-        /* Cancel the Timer task and free its resources */
-        vPortCancelThread( xTimerGetTimerDaemonTaskHandle() );
-    #endif /* configUSE_TIMERS */
+    // asking timer thread to shut down
+    timer_tick_thread_should_run = false;
+    pthread_join(timer_tick_thread, NULL);
+
+    // cancel and join any remaining pthreads
+    // to ensure their resources are freed
+    //
+    // https://stackoverflow.com/a/5612424
+    const size_t task_array_length = uxTaskGetNumberOfTasks();
+    TaskStatus_t pxTaskStatusArray[task_array_length];
+    configRUN_TIME_COUNTER_TYPE totalRunTime;
+    UBaseType_t ulNumThreads = uxTaskGetSystemState( pxTaskStatusArray,
```

Comment:
Can we consider to track the created tasks in the port instead of calling `uxTaskGetSystemState`? Then we don't have to depend on `configUSE_TRACE_FACILITY`.

For example,
```c
StackType_t * pxPortInitialiseStack( StackType_t * pxTopOfStack,
                                     StackType_t * pxEndOfStack,
                                     TaskFunction_t pxCode,
                                     void * pvParameters )
{
    /* Add the created thread to a list. */
}

void vPortCancelThread( void * pxTaskToDelete )
{
     /* Remove the thread from the list. */
}

BaseType_t xPortStartScheduler( void )
{
    ...
    while( xSchedulerEnd != pdTRUE )
    {
        sigwait( &xSignals, &iSignal );
    }

    /* Walk through the list to cancel all the threads here. */
    ...
}

```

Reply:
btw I had originally not done this in order to simplify the code but I see the list functions make it pretty simple to do.